### ISSUE TITLE
Limited length of name file using pthread + Side Module

### ISSUE THREAD

**Issue by henry0905** [emscripten-core/emscripten#14833]
Hello,

When creating a side module with the pthread option, an assertion occured during the loading of the side module **when** the side module's name exceed 16 characters.

The callstack is the following :
```
assert @ EVP.js:728
TextDecoderWrapper.decode @ EVP.js:850
UTF8ArrayToString @ EVP.js:877
getDylinkMetadata @ EVP.js:2317
receiveInstance @ EVP.js:1734
receiveInstantiationResult @ EVP.js:1768
```

The 16 characters limits seems to be defined inside UTF8ArrayToString :
```
  if (endPtr - idx > 16 && heap.subarray && UTF8Decoder) {
  return UTF8Decoder.decode(heap.subarray(idx, endPtr));
```

Here are the steps to reproduce (code of CPP files are below)

1. Case 1:  `-pthread` option + file name is over 16 characters: lead to assertion
Compilation:
```
REM Step 0: Remove the previous compilation
CALL rm *.html *.wasm *.js

REM Step 1: Create the binary object files with pthread option
CALL em++ -c foo2.cpp -pthread -fPIC -o foo2.o
CALL em++ -c -pthread -fPIC Main.cpp -o Main.o

REM Step 2: Create the wasm files with long name
CALL em++ foo2.o -s SIDE_MODULE=1 -s EXPORT_ALL=1 -pthread -fPIC -o qwertyuiopasdfghhjklcvbbcvbcv.wasm

REM Step 3: Final compilation
CALL em++ -pthread -s MAIN_MODULE=2 -s PROXY_TO_PTHREAD -s ALLOW_MEMORY_GROWTH=1 -fPIC -s EXIT_RUNTIME=1 Main.o -s RUNTIME_LINKED_LIBS="['qwertyuiopasdfghhjklcvbbcvbcv.wasm']" -o TEST.html
```
Result:
![image](https://user-images.githubusercontent.com/28931340/128880423-43d8c596-fdd3-49a3-8f89-4560fe6ad4fa.png)

2. Case 2: `-pthread` option + file name is less than 16 characters: Successful.
Compilation: (change from step 2)
```
REM Step 2: Create the wasm files and name has less than 16 characters
CALL em++ foo2.o -s SIDE_MODULE=1 -s EXPORT_ALL=1 -pthread -fPIC -o a.wasm

REM Step 3: Final compilation
CALL em++ -pthread -s MAIN_MODULE=2 -s PROXY_TO_PTHREAD -s ALLOW_MEMORY_GROWTH=1 -fPIC -s EXIT_RUNTIME=1 Main.o -s RUNTIME_LINKED_LIBS="['a.wasm']" -o TEST.html
```
Result:
![image](https://user-images.githubusercontent.com/28931340/128881773-edc4bf31-19cc-4ef5-87a3-1ccb30a46e76.png)

3. Case 3: NO `-pthread` option + file name is more than 16 characters: Successful.
Compilation: (change from step 1)
```
REM Step 1: Create the binary object files without -pthread option
CALL em++ -c foo2.cpp -fPIC -o foo2.o
CALL em++ -c -fPIC Main.cpp -o Main.o

REM Step 2: Create the wasm files with long name
CALL em++ foo2.o -s SIDE_MODULE=1 -s EXPORT_ALL=1 -fPIC -o qwertyuiopasdfghhjklcvbbcvbcv.wasm

REM Step 3: Final compilation
CALL em++ -s MAIN_MODULE=2 -fPIC -s EXIT_RUNTIME=1 Main.o -s RUNTIME_LINKED_LIBS="['qwertyuiopasdfghhjklcvbbcvbcv.wasm']" -o TEST.html
```
Result:
![image](https://user-images.githubusercontent.com/28931340/128882194-02ddfd71-6242-425e-9333-a2d119aca849.png)

My snippet code:
**Main.cpp**
```
#include <iostream>
#include <stdlib.h>

using namespace std;

int foo2();

int main()
{
	int result = foo2();
	printf("check foo2 %d\n", result);
	return 0;
}
```
**foo2.cpp**
```
int foo2()
{
    return 2;
}
```

**Comment by kripken** [comment]
Looks like the only assertion in that method is
```js
assert(data instanceof Uint8Array)
```
Can you add a `console.log(data)` right before it in your emitted JS file to print data, to see what type it is?

**Comment by henry0905** [comment]
The data type is Int8Array :
```js
Int8Array(34)
0: 113
1: 119
[...]
33: 109
buffer: (...)
byteLength: (...)
byteOffset: (...)
length: (...)
Symbol(Symbol.toStringTag): (...)
[[Prototype]]: TypedArray
```

After some investigation, data are initialized in the function `getDylinkMetadata`  :
```js
 if (binary instanceof WebAssembly.Module) {
  var dylinkSection = WebAssembly.Module.customSections(binary, "dylink");
  assert(dylinkSection.length != 0, "need dylink section");
  binary = new Int8Array(dylinkSection[0]);
 }
```

**Comment by kripken** [comment]
Thanks @henry0905 , that does look like the problem. Fix + test in #14854
